helpers.go

```diff
@@ -267,29 +268,111 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
+	for len(header) > 0 {
+		order++
+
+		// Skip spaces
+		header = utils.TrimLeft(header, ' ')
+
+		// Get spec
+		commaPos = strings.IndexByte(header, ',')
+		if commaPos != -1 {
+			spec = utils.Trim(header[:commaPos], ' ')
+		} else {
+			spec = utils.TrimLeft(header, ' ')
+		}
+
+		// Get quality
+		quality := 1.0
+		if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
+			factor := utils.Trim(spec[factorSign+1:], ' ')
+			if strings.HasPrefix(factor, "q=") {
+				if q, err := strconv.ParseFloat(factor[2:], 64); err == nil {
+					quality = q
+				}
+			}
+			spec = spec[:factorSign]
 		}
-		spec, commaPos := "", 0
-		for len(header) > 0 && commaPos != -1 {
-			commaPos = strings.IndexByte(header, ',')
+
+		// Skip if quality is 0.0
+		// See: https://www.rfc-editor.org/rfc/rfc9110#quality.values
+		if quality == 0.0 {
 			if commaPos != -1 {
-				spec = utils.Trim(header[:commaPos], ' ')
+				header = header[commaPos+1:]
 			} else {
-				spec = utils.TrimLeft(header, ' ')
-			}
-			if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
-				spec = spec[:factorSign]
+				break
 			}
+			continue
+		}
 
-			// isAccepted if the current offer is accepted
-			if isAccepted(spec, offer) {
-				return offer
+		// Get specificity
+		specificity := 0
+		// check for wildcard this could be a mime */* or a wildcard character *
+		if spec == "*/*" || spec == "*" {
+			specificity = 1
+		} else if strings.HasSuffix(spec, "/*") {
+			specificity = 2
+		} else if strings.IndexByte(spec, '/') != -1 {
+			specificity = 3
+		} else {
+			specificity = 4
+		}
+
+		// Add to accepted types
+		acceptedTypes = append(acceptedTypes, acceptedType{spec, quality, specificity, order})
+
+		// Next
+		if commaPos != -1 {
+			header = header[commaPos+1:]
+		} else {
+			break
+		}
+	}
+
+	// Sort accepted types by quality and specificity, preserving order of equal elements
+	if len(acceptedTypes) > 1 {
+		for i := 1; i < len(acceptedTypes); i++ {
+			swap := false
+			lo, hi := 0, i-1
+			for lo <= hi {
+				mid := (lo + hi) / 2
+				if acceptedTypes[i].quality < acceptedTypes[mid].quality ||
+					(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity < acceptedTypes[mid].specificity) ||
+					(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity == acceptedTypes[mid].specificity && acceptedTypes[i].order > acceptedTypes[mid].order) {
+					lo = mid + 1
+				} else {
+					hi = mid - 1
+				}
+			}
+			for j := i; j > lo; j-- {
+				acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
+				swap = true
+			}
+			if !swap {
+				break
 			}
+		}
+	}
 
-			if commaPos != -1 {
-				header = header[commaPos+1:]
+	// Find the first offer that matches the accepted types
+	for _, acceptedType := range acceptedTypes {
+		for _, offer := range offers {
+			if len(offer) == 0 {
+				continue
+			}
+			if isAccepted(acceptedType.spec, offer) {
+				return offer
```

Comment:
also, I have thought that. Does that make sense?
```suggestion
			if len(offer) == 0 || !isAccepted(acceptedType.spec, offer) {
				continue
			}
			return offer
```

Reply:
to me that's more compact but less redable, should make zero performance difference since it would likley product the same binary...